esp/bindings/http/platform/httptransport.cpp

```diff
@@ -2541,3 +2634,182 @@ void CHttpResponse::handleExceptions(IXslProcessor *xslp, IMultiException *me, c
+         || strnicmp(content_type, HTTP_TYPE_JAVASCRIPT, strlen(HTTP_TYPE_JAVASCRIPT)) == 0
+         || strnicmp(content_type, HTTP_TYPE_JAVASCRIPT2, strlen(HTTP_TYPE_JAVASCRIPT2)) == 0))
+        return false;
+
+    return true;
+#endif
+}
+
+bool CHttpResponse::compressContent(StringBuffer* originalContent, int compressType)
+{
+#ifdef _USE_ZLIB
+    ZlibCompressionType zlibtype = static_cast<ZlibCompressionType>(compressType);
+    MemoryBuffer zippedContent;
+    try
+    {
+        zlib_deflate(zippedContent, m_content.str(), m_content.length(), GZ_DEFAULT_COMPRESSION, zlibtype);
+    }
+    catch(IException* e)
+    {
+        StringBuffer estr;
+        IERRLOG("Exception(%d, %s) compressing response content.", e->errorCode(), e->errorMessage(estr).str());
+        return false;
+    }
+    catch(...)
+    {
+        IERRLOG("Unknown exception compressing response content.");
+        return false;
+    }
+    if (originalContent != nullptr)
+        originalContent->setown(m_content);
+    setContent(zippedContent.length(), zippedContent.toByteArray());
+    addHeader(HTTP_HEADER_CONTENT_ENCODING, zlibType2Header(zlibtype));
+    return true;
+#endif
+}
+
+bool CHttpResponse::shouldDecompress(int& compressType)
+{
+    compressType = 0;
+    if (!m_compressionEnabled)
+        return false;
+    if (m_content_length == 0 || m_content.length() == 0)
+        return false;
+
+#ifdef _USE_ZLIB
+    StringBuffer ceheader;
+    getHeader(HTTP_HEADER_CONTENT_ENCODING, ceheader);
+    ceheader.trim();
+    if (ceheader.length() == 0)
+        return false;
+    if (strieq(ceheader.str(), "gzip"))
+        compressType = static_cast<int>(ZlibCompressionType::GZIP);
+    else if (strieq(ceheader.str(), "deflate"))
+        compressType = static_cast<int>(ZlibCompressionType::DEFLATE);
+    else if (strieq(ceheader.str(), "x-deflate"))
+        compressType = static_cast<int>(ZlibCompressionType::ZLIB_DEFLATE);
+    else
+        return false;
+    return true;
+#else
+    return false;
+#endif
+}
+
+bool CHttpResponse::decompressContent(StringBuffer* originalContent, int compressType)
+{
+#ifdef _USE_ZLIB
+    ZlibCompressionType zlibtype = static_cast<ZlibCompressionType>(compressType);
+    StringBuffer decompressedContent;
+    try
+    {
+        httpInflate((const unsigned char*)m_content.str(), m_content.length(), decompressedContent, zlibtype==ZlibCompressionType::GZIP);
+    }
+    catch(IException* e)
+    {
+        StringBuffer estr;
+        IERRLOG("Exception(%d, %s) decompressing response content.", e->errorCode(), e->errorMessage(estr).str());
+        return false;
+    }
+    catch(...)
+    {
+        IERRLOG("Unknown exception decompressing response content.");
+        return false;
+    }
+    if (originalContent != nullptr)
+        originalContent->setown(m_content);
+    m_content.setown(decompressedContent);
+    m_content_length = m_content.length();
+    removeHeader(HTTP_HEADER_CONTENT_ENCODING);
```

Comment:
@wangkx it's a close call, but I think it's slightly better/safer to remove the header because (1) the whole message has been transformed which includes content & content_length and (2) to avoid upper layers to do anything extra with the encoding header